---
title: "PanomiR Package Walkthrough"
output: 
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{PanomiR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

PanomiR is a package for pathway and microRNA Analysis of gene expression data.
This document provides details about how to install and utilize various 
functionality in PanomiR

## Installation

PanomiR can be accessed via Bioconductor. To install, start R (version > "4.1)
and run the following code.

```{r installation_bioc, eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("PanomiR")
```

You can also install the latest development version of PanomiR using GitHub.

```{r installation_git, eval = FALSE}
devtools::install_github("pouryany/PanomiR")
```


## 1. Pathway summarization

PanomiR can generate pathway activity profiles given a gene expression dataset
and a list of pathways. This section uses an example dataset from the Cancer
Genome Atlas (TCGA) Liver Hepatocellular Carcinoma (LIHC) dataset to generate
Pathway summary statistics. 

```{r load_package}
library(PanomiR)

# Downloading example LIHC expression dataset
LIHC_url  <- url(paste0("https://github.com/pouryany/PanomiR_paper",
                        "/raw/main/data/TCGA_LIHC.RDS"))
TCGA_LIHC <- readRDS(LIHC_url)

# Downloading example LIHC covariates dataset
LIHC_cov_url  <- url(paste0("https://github.com/pouryany/PanomiR_paper",
                        "/raw/main/data/covariates_TCGA_LIHC.RDS"))
cov_TCGA_LIHC <- readRDS(LIHC_cov_url)

# Pathway reference from the PanomiR package
data("path_gene_table")

# Generating pathway summary statistics 

summaries <- pathwaySummary(TCGA_LIHC,path_gene_table,method = "x2")

head(summaries)[,1:2]
```


## 2. Differential Pathway activation

Once you generate the pathway activity profiles, as discussed in the last
section, there are several analysis that you can perform. We have bundled some 
of the most important ones into standalone functions. Here, we describe
differential pathway activation profiling, which is examining differenes in
pathway activity profiles in user-determined conditions.


```{r differential}

output0 <- differentialPathwayAnalysis(geneCounts = TCGA_LIHC,
                                       pathways =  path_gene_table,
                                       covariates = cov_TCGA_LIHC, 
                                       condition = 'shortLetterCode',
                                       adjustCovars ='plate')

de.paths <- output0$DEP

head(de.paths)
```

## 3. Finding clusters of pathways

PanomiR provides a function to find groups coordinated differentially 
activated pathways based on the PCxN methodology. 

```{r pcxn}


# using an updated version of pcxn 
pcxn_url  <- url(paste0("https://github.com/pouryany/PanomiR_paper",
                        "/raw/main/data/pcxn_panomir.RDS"))
pcxn_net  <- readRDS(pcxn_url)

set.seed(2)
pathwayClustsLIHC <- mappingPathwaysClusters(pcxn = pcxn_net, 
                            dePathways = de.paths[1:300,],
                            topPathways = 200,
                            outDir=".",
                            plot = FALSE,
                            subplot = FALSE,
                            prefix='',
                            clusteringFunction = "cluster_louvain",
                            correlationCutOff = 0.1)


head(pathwayClustsLIHC$Clustering)

```


## 4. Prioritizing miRNAs per cluster of pathways.

PanomiR identifies miRNAs that target clusters of pathways, as defined in the 
last section. In order to this, you would need a reference table of
miRNA-Pathway association score (enrichment). We recommend using a customized
miRNA-Pathway association table, tailored to your experimental data. Here, 
we provide a pre-processed table for LIHC table and in the next section, we will
explain how to generate the customized tables.

Note that in the example below, we use a sampling rate of 50, the recommended
rate is between 500-1000. Also, we set the saveSampling argument to FALSE. This
argument should be set to TRUE if you wish to save your sampling and check for
different outputs from the clustering algorithms or pathway thresholds.


```{r miRNA}


# using an updated version of pcxn 
enrch_url    <- url(paste0("https://github.com/pouryany/PanomiR_paper",
                        "/raw/main/data/LIHC_ENRICHMENT_TargetScan03.RDS"))
tableEnrich  <- readRDS(enrch_url)

set.seed(1)
output2 <- prioritizeMicroRNA(enriches0 = tableEnrich,
                              pathwayClusters = pathwayClustsLIHC$Clustering,
                              topClust = 1,
                              sampRate = 50, 
                              method = c("aggInv"),
                              outDir = "Output/",
                              dataDir = "outData/",
                              saveSampling = FALSE,
                              runJackKnife = FALSE,
                              numCores = 1,
                              prefix = "outmiR",
                              saveCSV = FALSE)

head(output2$Cluster1)

```



## 5. miRNA-Pathway enrichment tables

PanomiR best performs on tissue/experiment-customized datasets. In order to do
this, you need to create a customized enrichment table. You can simply do so by
using the pathway and miRNA list that we have provided as a part of the package.
simply, plug in the name of the genes present (expressed) in your experiment in
the following code



```{r customized_mir, eval=FALSE}


# using an updated version of pcxn 
data("msigdb_c2")
data("targetScan_03")


customeTableEnrich <- miRNAPathwayEnrichment(mirSets = targetScan_03,
                                              pathwaySets = msigdb_c2,
                                              geneSelection = yourGenes,
                                              mirSelection = yourMicroRNAs,
                                              fromID = "ENSEMBL",
                                              toID = "ENTREZID",
                                              minPathSize = 9,
                                              numCores = 1,
                                              outDir = ".",
                                              saveOutName = NULL)

```

In the above section, the field ```fromID``` denotes the gene representation 
format of your input list. Here is a quick example that runs fast.
```{r customized_mir2, eval=FALSE}


# using an updated version of pcxn 
data("msigdb_c2")
data("targetScan_03")

tempEnrich <-miRNAPathwayEnrichment(targetScan_03[1:20],msigdb_c2[1:20])

head(tempEnrich)
```

## Session info
```{r sessionInfo}
sessionInfo()

```
